File: lua/src/loops.lua
-- Module for demonstrating loops in Lua
local M = {}

function M.demonstrate_loops()
    print("=== Loops ===")
    
    -- Numeric for loop
    print("Numeric for loop:")
    for i = 1, 5 do
        io.write(i, " ")
    end
    print()
    
    -- Numeric for loop with step
    print("\nNumeric for loop with step of 2:")
    for i = 0, 10, 2 do
        io.write(i, " ")
    end
    print()
    
    -- Generic for loop with ipairs (array-like iteration)
    print("\nGeneric for loop with ipairs:")
    local fruits = {"apple", "banana", "orange"}
    for i, fruit in ipairs(fruits) do
        print(i, fruit)
    end
    
    -- Generic for loop with pairs (table iteration)
    print("\nGeneric for loop with pairs:")
    local person = {
        name = "Alice",
        age = 25,
        city = "Wonderland"
    }
    for key, value in pairs(person) do
        print(key, value)
    end
    
    -- While loop
    print("\nWhile loop:")
    local counter = 1
    while counter <= 5 do
        print("Counter:", counter)
        counter = counter + 1
    end
    
    -- Repeat-until loop
    print("\nRepeat-until loop:")
    local x = 1
    repeat
        print("x:", x)
        x = x + 1
    until x > 5
    
    -- Breaking out of loops
    print("\nLoop with break:")
    for i = 1, 10 do
        if i > 5 then
            break
        end
        io.write(i, " ")
    end
    print()
    
    -- Using goto for complex flow control
    print("\nUsing goto:")
    local count = 1
    ::continue::
    if count <= 3 then
        print("Count:", count)
        count = count + 1
        goto continue
    end
    
    print() -- Empty line
end

return M

File: ruby/lib/loops.rb
module Loops
  def self.demonstrate
    puts "\n=== Loops ==="
    
    # for loop with range
    print "For loop with range: "
    for i in 1..4
      print "#{i} "
    end
    puts
    
    # each iterator (more Ruby-like)
    print "Each iterator: "
    (1..4).each { |i| print "#{i} " }
    puts
    
    # while loop
    print "\nWhile loop: "
    count = 1
    while count < 5
      print "#{count} "
      count += 1
    end
    puts
    
    # until loop (opposite of while)
    print "Until loop: "
    count = 1
    until count >= 5
      print "#{count} "
      count += 1
    end
    puts
    
    # times iterator
    print "\n5 times: "
    5.times { |i| print "#{i} " }
    puts
    
    # upto/downto
    print "Up to: "
    1.upto(4) { |i| print "#{i} " }
    puts
    
    print "Down to: "
    4.downto(1) { |i| print "#{i} " }
    puts
    
    # loop do (with break)
    print "\nLoop do with break: "
    count = 1
    loop do
      print "#{count} "
      count += 1
      break if count >= 5
    end
    puts
    
    # each_with_index
    puts "\nEach with index:"
    fruits = ["apple", "banana", "orange"]
    fruits.each_with_index do |fruit, index|
      puts "#{index + 1}. #{fruit}"
    end
    
    # next (continue) and redo
    puts "\nUsing next and redo:"
    5.times do |i|
      next if i == 2  # Skip 2
      print "#{i} "
    end
    puts
  end
end